enhancements/installer/agent-late-binding.md

```diff
@@ -0,0 +1,179 @@
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+In the current implementation of Assisted Installer, each Discovery ISO is
+associated with an existing cluster resource (a cluster definition that may not
+yet be an installed OCP cluster).  When an agent comes up, it registers with
+the Assisted Service and is automatically associated with that same cluster
+resource.  The only way to associate that host with a different cluster is to
+boot it with an ISO associated with the desired cluster.  We can see that the
+process of configuring the infrastructure is closely tied to cluster creation.
+
+The late binding feature splits the process into two parts, each potentially
+performed by a different persona.  The Discovery ISO is created by an
+infrastructure admin, who then boots the hosts.  Each host becomes visible to
+the cluster creators but not associated with any specific cluster resource.  At
+a later time, a cluster creator can create or expand a cluster using available
+hosts.
+
+## Motivation
+
+The benefits of late binding are as follows:
+- Adjusts the flow to suit the two different personas (infrastructure
+  administrator and cloud creator).
+- Allows for a more cloud-like use case where compute resources of different
+  flavors can be consumed, similar to IBM Satellite.
+- Allows agent-based installation to fit more naturally into a MAPI or CAPI
+  model.  While the details of such an integration are not in the scope of
+  this enhancement, the idea is that hosts are booted, thus creating a
+  collection of unassigned agents, and the machine creation involves choosing one
+  such agent and beginning its installation.
+- Reduces the number of Discovery ISOs, as there is no need to generate and
+  download one for each cluster installation.  Each ISO incurs capacity
+  overhead for the service, and management overhead for the user which conflicts
+  with our ease-of-use goal.
+- Allows the user to see their inventory and then make decisions on how to
+  build clusters based on hardware and connectivity.  At a later stage, the
+  service can make recommendations.  This would be a layer of logic above late
+  binding and is out of scope for this enhancement.
+- Enables the user story where the Discovery ISO is loaded at the factory, and
+  then discovered at a later time.
+
+### Goals
+
+- Make the InfraEnv and Agent resources independent from the ClusterDeployment.
+  A single InfraEnv should have the potential to boot any number of hosts,
+  whose Agents register with no relation to any particular ClusterDeployment.  An
+  Agent may be assigned to a ClusterDeployment at a later time.
+- An Agent may be associated with a different ClusterDeployment at any time.
```

Comment:
"any time" is a lot of times. :) Is this true after installation has started for example?